### ISSUE TITLE
Add support for merging jars when package are split across jars

### ISSUE THREAD

**Issue by thekalinga** [jjohannes/extra-java-module-info#1]
Currently there seems to be no way to specify multiple jars to be specified to be merged into a single jar when the packages are split across. When such a situation is encountered, currently AFAIK, we dont have any choice but to manually fiddle.

Can you add support for specifying multiple jars and specify the combined exported packages for all of them in one shot?

**Comment by jjohannes** [comment]
@thekalinga that's and interesting use case, because it can not be solved the way this plugin currently works. Gradle treats each "module" that has its own coordinates in a repository and the artifacts (jars) behind it independently. So the _artifact transform_ that is provided by this plugin operates on each Jar independently.

To merge two Jars that are located at different coordinates, and are therefore treated as two completely independent things, would require another solution.

Can you share a concrete use cases with Jars published to public repositories that have this issue?

**Comment by thekalinga** [comment]
Sorry for the delay in responding.

Here is the repo with a library that's used to read epub files & the issues we encounter when using legacy jars

1. Merging without overriding (should be the default as this helps in debugging & forces people to explicitly specify override option explicitly)
2. Merging with overriding contents (as packages are split across multiple modules), in this option, the order matters

https://github.com/thekalinga/gradle-merge-modules-usecase

![image](https://user-images.githubusercontent.com/452518/93385145-7117a280-f883-11ea-96b8-a7af452ad015.png)

As you can see the library `xmlpull` exposes `org.xmlpull.v1` packages with just 2 classes & `org.xmlpull.mxp1` that are unique to this jar, where as `kxml2` exposes `org.kxml2` & `org.xmlpull.v1` which has 4 classes

The way I fixed tried fixing this temporarily was by using this elaborate dance & its quite awful

```groovy
automaticModule("epublib-core-${Constants.epubLibVersion}.jar", "nl.siegmann.epublib")
automaticModule("xmlpull-1.1.3.4d_b4_min.jar", "org.xmlpull")
automaticModule("kxml2-${Constants.kxmlVersion}.jar", "net.sf.kxml")
module("xmlpull-1.1.3.4d_b4_min.jar", "org.xmlpull", "1.1.3.4d_b4_min") {
        exports("org.xmlpull.mxp1")
}
```

As of now, I am not even using this fix, rather I am using some other unofficial libraries of `kxml2` & `xmlpull` which claim compatibility. Its more or less trial and error.

Legacy libraries require merging

**Comment by iherasymenko** [comment]
@thekalinga 

Looking at the two jars in question, it seems like you can safely solve your problem by substituting one dependency with another utilizing the power of Gradle's dependency management subsystem.

```
configurations.all { Configuration configuration ->
    configuration.resolutionStrategy.dependencySubstitution {    
        substitute module('xmlpull:xmlpull') because 'it is a superset of net.sf.kxml:kxml2' with module('net.sf.kxml:kxml2:2.3.0')
    }
}
```

`org.xmlpull.v1.XmlPullParserException` and `org.xmlpull.v1.XmlPullParserException` are identical in both jars, while `org.xmlpull.mxp1.MXParser` from `xmlpull:xmlpull` is an SPI implementation registered via Java ServiceLoader. Its counterpart `net.sf.kxml:kxml2:2.3.0` provides another implementation (`org.kxml2.io.KXmlParser`) accordingly.

As long as there are no hard references to `org.xmlpull.mxp1.MXParser`, you should be good.

**Comment by thekalinga** [comment]
@iherasymenko Thanks for showing me another way (I've not verified references, but I assume you are right).

As you can understand, the point of the issue is not just about this, but this is something I came across in my limited cases.

And due to the way this is replaced, people cant use anything under `org.xmlpull.mxp1` within their code/any library in the classpath, which is a significant limitation

Probably we can wait till someone actually comes up with a scenario where they has to be fixed, but I suspect this is an issue that will happen sooner (or) later

**Comment by jjohannes** [comment]
Thanks for the additional explanations and the sample project. I see how there can be certain use cases for such a "merge" functionality. I think this would need to be done by a task the pre-processes the classpath before compilation. I'll try something.

**Comment by jjohannes** [comment]
I have been looking into this and I have spike implementation for defining Jar merges which are then done by a task that pre-processes the classpath. Which is cool (if it works) but also more involved. Because this would have to be wired in in several places (compile task, run task, ...).

I started writing a test, but I begin wondering about what the real use cases are here.

First the use case shown above (with xmlpull) does not seem to be a real one as @iherasymenko pointed out. You should not have both Jars on the classpath and that this gave an error is actually a good thing from that perspective.

Then @thekalinga wrote that he could solve the issue by something like this:
```
module("xmlpull-1.1.3.4d_b4_min.jar", "org.xmlpull", "1.1.3.4d_b4_min") {
        exports("org.xmlpull.mxp1")
}
```

Which is exactly what this plugin is about: To allow you to define complete module descriptors that only give limited visibility to packages, which can also help to avoid certain "split" problems.

Does anyone have some "real" use cases with which I could test the implementation I am working on?

**Comment by iherasymenko** [comment]
@jjohannes `org.slf4j:slf4j-api:1.7.30` and `org.slf4j:slf4j-ext:1.7.30` suffer from the split package problem. Both JARs have classes in `org.slf4j` package, therefore they cannot be placed on the module path. I think this would be a relevant example. The issue will be fixed in SLF4J 2, however it is still in alpha.

**Comment by Burtan** [comment]
[ObjectBox has the same problems.](https://github.com/objectbox/objectbox-java)

**Comment by jjohannes** [comment]
Thanks for the pointers. I am still interested in exploring this, but didn't have much time recently to work on it. I pushed what I have so far here: https://github.com/jjohannes/extra-java-module-info/tree/merge-jars

I tried it with the `slf4j` case like this:

```
extraJavaModuleInfo {
    automaticModule("slf4j-api-1.7.30.jar", "org.slf4j") {
        mergeJar("slf4j-ext-1.7.30.jar")
    }
}

dependencies {
    implementation("org.slf4j:slf4j-api:1.7.30")
    implementation("org.slf4j:slf4j-ext:1.7.30")
}
```

In general it already seems to work. Just the naive Jar merging I have is not working out because, I think, classes from the same package need to be ordered correctly. If anyone feels like picking this up and improve this, please feel free to do so.

In case you don't know, you can work with the development version of the plugin in your project:
- Clone this repo
- Switch to branch
- add `includeBuild("path/to/repo")` in `settings.gradle(.kts)`

**Comment by jjohannes** [comment]
Finally, I got back to this and got a (I think) much better solution done that works directly in the _Transform_ - see 62957d76fc61ee724a15483ca2b8ad624788e2d4 and 329293473e9bd22417176f2c68293b66608a3e38.

This is now available int Version `0.12`. See also docs here:
https://github.com/jjohannes/extra-java-module-info#what-do-i-do-in-a-split-package-situation

Thanks everyone for the discussion and feedback. Let me know if there are issues with the solution.
